subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -926,3 +1031,103 @@ int nrf_cloud_fota_mqtt_evt_handler(const struct mqtt_evt *evt)
 
 	return 0;
 }
+
+#if IS_ENABLED(CONFIG_NRF_CLOUD_FOTA_BLE_DEVICES)
+int nrf_cloud_fota_ble_set_handler(nrf_cloud_fota_ble_callback_t cb)
+{
+	if (!cb) {
+		return -EINVAL;
+	}
+
+	ble_cb = cb;
+	return 0;
+}
+
+
+int nrf_cloud_fota_ble_update_check(const bt_addr_t * const ble_id)
+{
+	if (ble_id == NULL) {
+		return -EINVAL;
+	} else if (client_mqtt == NULL) {
+		return -ENXIO;
+	} else if (topic_req.topic.utf8 == NULL) {
+		return -EADDRNOTAVAIL;
+	}
+
+	char payload[BLE_JOB_REQUEST_PAYLOAD_SIZE];
+	char ble_id_str[BT_ADDR_LE_STR_LEN];
+	struct mqtt_publish_param param = {
+		.message_id = NRF_CLOUD_FOTA_BLE_REQUEST_ID,
+		.dup_flag = 0,
+		.retain_flag = 0,
+	};
+
+	int ret = bt_addr_to_str(ble_id, ble_id_str, BT_ADDR_LE_STR_LEN);
+
+	if (ret != sizeof(ble_id->val)) {
+		return -EADDRNOTAVAIL;
+	}
+
+	ret = snprintf(payload, BLE_JOB_REQUEST_PAYLOAD_SIZE,
+		       BLE_JOB_REQUEST_LATEST_TEMPLATE,
+		       ble_id_str);
+	if (ret <= 0 || ret >= BLE_JOB_REQUEST_PAYLOAD_SIZE) {
+		return -E2BIG;
+	}
+
+	param.message.topic = topic_ble_req;
+	param.message.payload.data = payload;
+	param.message.payload.len = ret;
+
+	return publish(&param);
+}
+
+int nrf_cloud_fota_ble_job_update(const struct nrf_cloud_fota_ble_job
+				  * const ble_job,
+				  const enum nrf_cloud_fota_status status)
+{
+	if (ble_job == NULL) {
+		return -EINVAL;
+	} else if (topic_ble_updt.topic.utf8 == NULL) {
+		return -ENODEV;
+	} else if (client_mqtt == NULL) {
+		return -ENXIO;
+	}
+
+	int ret;
+	char payload[UPDATE_PAYLOAD_SIZE];
+	char ble_id_str[BT_ADDR_LE_STR_LEN];
+	struct mqtt_publish_param param = {
+		.message_id = NRF_CLOUD_FOTA_BLE_UPDATE_ID,
+		.dup_flag = 0,
+		.retain_flag = 0,
+	};
+
+	ret = bt_addr_to_str(&ble_job->ble_id, ble_id_str, BT_ADDR_LE_STR_LEN);
+	if (ret < 0 || ret >= BT_ADDR_LE_STR_LEN) {
+		return -EADDRNOTAVAIL;
+	}
+
+	if (status == NRF_FOTA_DOWNLOADING) {
+		ret = snprintf(payload, UPDATE_PAYLOAD_SIZE,
+			 BLE_JOB_UPDATE_PROGRESS_TEMPLATE,
+			 ble_id_str, ble_job->info.id,
+			 status, ble_job->dl_progress);
+	} else {
+		ret = snprintf(payload, UPDATE_PAYLOAD_SIZE,
+			 BLE_JOB_UPDATE_MSG_TEMPLATE,
+			 ble_id_str, ble_job->info.id,
+			 status, get_error_string(ble_job->error));
```

Comment:
I'm thinking it might be helpful to have a BLE-specific error to indicate the device has gone offline unexpectedly, during the BLE DFU.

Reply:
agreed.  we can add error codes to the enum as needed.